magma/fromverilog.py

```diff
@@ -111,48 +131,50 @@ def ParseVerilogModule(node, type_map):
 
     return node.name, args
 
-def FromVerilog(source, func, type_map, target_modules=None):
-    parser = VerilogParser()
 
+def FromVerilog(source, func, type_map, target_modules=None, shallow=False):
+    parser = VerilogParser()
     ast = parser.parse(source)
-    #ast.show()
-
-    v = ModuleVisitor()
-    v.visit(ast)
-
-    if func == DefineCircuit:
-        # Only allow a single verilog module unless we're only defining one
-        # circuit (only one module in target_modules), otherwise, they would all
-        # use the same source, so if they are compiled together, there will be
-        # multiple definitions of the same verilog module.
-        assert len(v.nodes) == 1 or (target_modules and len(target_modules) == 1)
-    modules = []
-    for node in v.nodes:
-        if target_modules is not None and node.name not in target_modules:
-            continue
-        try:
-            name, args = ParseVerilogModule(node, type_map)
-            circuit = func(name, *args)
-            if func == DefineCircuit:
-                # inline source
-                circuit.verilogFile = source
-                EndDefine()
-            circuit.verilog_source = source
-            modules.append(circuit)
-        except Exception as e:
-            logger.warning(f"Could not parse module {node.name} ({e}), "
-                           f"skipping")
-    if not modules:
+    visitor = ModuleVisitor()
+    visitor.visit(ast)
+
+    def _get_lines(start_line, end_line):
+        if shallow:
+            return source
+        lines = source.split("\n")
+        return "\n".join(lines[start_line - 1:end_line])
+
+    magma_defns = {}
+    for name, verilog_defn in visitor.defns.items():
+        parsed_name, args = ParseVerilogModule(verilog_defn, type_map)
+        assert parsed_name == name
+        magma_defn = func(name, *args)
+        if func == DefineCircuit:
+            # Attach relevant lines of verilog source.
+            magma_defn.verilogFile = _get_lines(
+                verilog_defn.lineno, verilog_defn.end_lineno)
+            if not shallow:
+                for instance in visitor.get_instances(verilog_defn):
+                    instance_defn = magma_defns[instance.module]
+                    instance_defn()
+            EndDefine()
+        magma_defn.verilog_source = source
+        magma_defns[name] = magma_defn
```

Comment:
I think we shouldn't support this. The reason I got rid of the try/catch is it can potentially hide real errors, and mask them as "bad parsing". We can add something back in after, but we should catch specific errors and report them IMO.